cmake/external/mkldnn.cmake

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2017 PaddlePaddle Authors. All Rights Reserve.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+IF(NOT ${WITH_MKLDNN})
+  return()
+ENDIF(NOT ${WITH_MKLDNN})
+
+INCLUDE(ExternalProject)
+
+SET(MKLDNN_PROJECT "extern_mkldnn")
+SET(MKLDNN_SOURCES_DIR ${THIRD_PARTY_PATH}/mkldnn)
+SET(MKLDNN_INSTALL_DIR ${THIRD_PARTY_PATH}/install/mkldnn)
+SET(MKLDNN_INCLUDE_DIR "${MKLDNN_INSTALL_DIR}/include" CACHE PATH "mkldnn include directory." FORCE)
+
+# The following magic numbers should be updated regularly to keep latest version
+SET(MKLDNN_TAG "v0.9")
+SET(MKLDNN_MKL_VER "mklml_lnx_2018.0.20170425")
+
+IF(WIN32)
+    MESSAGE(WARNING "It is not supported compiling with mkldnn in windows Paddle yet."
+      "Force WITH_MKLDNN=OFF")
+    SET(WITH_MKLDNN OFF)
+    return()
+ELSE(WIN32)
+    SET(MKLDNN_LIBRARY "${MKLDNN_INSTALL_DIR}/lib/libmkldnn.so" CACHE FILEPATH "mkldnn library." FORCE)
+    MESSAGE(STATUS "Set ${MKLDNN_INSTALL_DIR}/lib to runtime path")
+    SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+    #SET(CMAKE_MACOSX_RPATH 1) # hold for MacOS
+    SET(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_RPATH}" "${MKLDNN_INSTALL_DIR}/lib")
+ENDIF(WIN32)
+
+INCLUDE_DIRECTORIES(${MKLDNN_INCLUDE_DIR})
+
+SET(MKLDNN_CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+SET(MKLDNN_CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+
+ExternalProject_Add(
+    ${MKLDNN_PROJECT}
+    ${EXTERNAL_PROJECT_LOG_ARGS}
+    GIT_REPOSITORY    "https://github.com/01org/mkl-dnn.git"
+    GIT_TAG           "${MKLDNN_TAG}"
+    PREFIX            ${MKLDNN_SOURCES_DIR}
+    PATCH_COMMAND     cd <SOURCE_DIR>/scripts && ./prepare_mkl.sh
```

Comment:
In my server, I met the following when downloading `mklml_lnx_2018.0.20170425.tgz`. I see it succeeded in Travis CI. I'm not sure why.
```
$ wget -P ../external/ https://github.com/01org/mkl-dnn/releases/download/v0.9/mklml_lnx_2018.0.20170425.tgz
--2017-07-18 10:20:14--  https://github.com/01org/mkl-dnn/releases/download/v0.9/mklml_lnx_2018.0.20170425.tgz
Resolving agent.baidu.com... 172.22.15.16, 172.22.15.17
Connecting to agent.baidu.com|172.22.15.16|:8128... connected.
Proxy request sent, awaiting response... 302 Found
Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102015Z&X-Amz-Expires=300&X-Amz-Signature=54ec18736c48ae3fcf97dea15db01075d8ef50171c54ab396325494eaba57a9f&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream [following]
--2017-07-18 10:20:15--  https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102015Z&X-Amz-Expires=300&X-Amz-Signature=54ec18736c48ae3fcf97dea15db01075d8ef50171c54ab396325494eaba57a9f&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream
Connecting to agent.baidu.com|172.22.15.16|:8128... connected.
Proxy request sent, awaiting response... 200 OK
Length: 67379285 (64M) [application/octet-stream]
../external/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102015Z&X-Amz-Expires=300&X-Amz-Signature=54ec18736c48ae3fcf97dea15db01075d8ef50171c54ab396325494eaba57a9f&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment; filename=mklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream: File name too long

Cannot write to `../external/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102015Z&X-Amz-Expires=300&X-Amz-Signature=54ec18736c48ae3fcf97dea15db01075d8ef50171c54ab396325494eaba57a9f&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment; filename=mklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream' (No such file or directory).
```

However, I can successfully download using the following command:
```
$ wget -O ../external/mklml_lnx_2018.0.20170425.tgz https://github.com/01org/mkl-dnn/releases/download/v0.9/mklml_lnx_2018.0.20170425.tgz 
--2017-07-18 10:20:45--  https://github.com/01org/mkl-dnn/releases/download/v0.9/mklml_lnx_2018.0.20170425.tgz
Resolving agent.baidu.com... 172.22.15.17, 172.22.15.16
Connecting to agent.baidu.com|172.22.15.17|:8128... connected.
Proxy request sent, awaiting response... 302 Found
Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102046Z&X-Amz-Expires=300&X-Amz-Signature=32ff43f10569d40b3fb22e5237de554764056e53b7397393aeb993bf92754f9d&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream [following]
--2017-07-18 10:20:46--  https://github-production-release-asset-2e65be.s3.amazonaws.com/58414589/54f3715e-3c9a-11e7-97f0-b402ecc00a80?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20170718%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170718T102046Z&X-Amz-Expires=300&X-Amz-Signature=32ff43f10569d40b3fb22e5237de554764056e53b7397393aeb993bf92754f9d&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dmklml_lnx_2018.0.20170425.tgz&response-content-type=application%2Foctet-stream
Connecting to agent.baidu.com|172.22.15.17|:8128... connected.
Proxy request sent, awaiting response... 200 OK
Length: 67379285 (64M) [application/octet-stream]
Saving to: `../external/mklml_lnx_2018.0.20170425.tgz'

100%[===================================================================================================================================================>] 67,379,285  1.03M/s   in 67s     

2017-07-18 10:21:54 (977 KB/s) - `../external/mklml_lnx_2018.0.20170425.tgz' saved [67379285/67379285]
```

Besides, I suggest adding another cmake file, such as `cmake/external/mkl-lite.cmake`, and using another external command to download `mklml_lnx_2018.0.20170425.tgz`, as it can be accessed independently. @hedaoyuan @luotao1 , what's your opinion?